agent/lib/kontena/network_adapters/weave.rb

```diff
@@ -76,25 +86,81 @@ def already_started?
       @started == true
     end
 
+    def cleanup_ipam
+      loop do
+        sleep 60 * 3 # 3mins
+        debug "starting ipam cleanup process"
+        ipam_container = Docker::Container.get('kontena-ipam-plugin') rescue nil
+        if ipam_container
+          local_addresses = []
+          Docker::Container.all(all: true).each do |container|
+            local_addresses << container.overlay_cidr
+          end
+          cmd = ['/app/bin/kontena-ipam-cleanup'] + local_addresses.compact
+          debug "executing cleanup with command: #{cmd}"
+          ipam_container.exec(cmd) { |stream, chunk| debug "#{stream}: #{chunk}" }
+        end
+      end
+    end
+
     # @param [Hash] opts
     def modify_create_opts(opts)
+      ensure_weave_wait
+
+      image = Docker::Image.get(opts['Image'])
+      image_config = image.info['Config']
+      cmd = []
+      if opts['Entrypoint']
+        if opts['Entrypoint'].is_a?(Array)
+          cmd = cmd + opts['Entrypoint']
+        else
+          cmd = cmd + [opts['Entrypoint']]
+        end
+      end
+      if !opts['Entrypoint'] && image_config['Entrypoint'] && image_config['Entrypoint'].size > 0
+        cmd = cmd + image_config['Entrypoint']
+      end
+      if opts['Cmd'] && opts['Cmd'].size > 0
+        if opts['Cmd'].is_a?(Array)
+          cmd = cmd + opts['Cmd']
+        else
+          cmd = cmd + [opts['Cmd']]
+        end
+      elsif image_config['Cmd'] && image_config['Cmd'].size > 0
+        cmd = cmd + image_config['Cmd']
+      end
+      opts['Entrypoint'] = ['/w/w']
+      opts['Cmd'] = cmd
+
       modify_host_config(opts)
+      opts
+    end
+
+    # @param [Hash] opts
+    def modify_network_opts(opts)
+      opts['Labels']['io.kontena.container.overlay_cidr'] = @ipam_client.reserve_address('kontena')
+      opts['Labels']['io.kontena.container.overlay_network'] = 'kontena'
 
       opts
     end
 
     # @param [Hash] opts
     def modify_host_config(opts)
       host_config = opts['HostConfig'] || {}
+      host_config['VolumesFrom'] ||= []
+      host_config['VolumesFrom'] << "weavewait-#{WEAVE_VERSION}:ro"
       dns = interface_ip('docker0')
       if dns && host_config['NetworkMode'].to_s != 'host'
         host_config['Dns'] = [dns]
         host_config['DnsSearch'] = ['kontena.local']
+        # TODO Do we still need TCP mode for plain weave networking?
         host_config['DnsOptions'] = ['use-vc'] # tcp mode for dns lookups
```

Comment:
@jakolehm @SpComb Do we still want to keep TCP mode, or switch back to default UDP mode?